### PR TITLE
imxrt: Ethernet Cleanup and extend support for ETH1 or ETH2 and LAN8742A

### PR DESCRIPTION
**Pull request by davids5**
## Summary

- Added Support for LAN8742A PHY
- All boards ARCH_PHY_INTERRUPT is a board property
- add IMXRT_PHY_POLLING
- Support either ETH1 or ETH2
- pinmux Add ENET2 and correct ALT for GPIO_ENET2_REF_CLK2
- Support WB dcache:  ensure proper dcache operation in Writeback mode
-  Better interrupt state Handling: Proper order of operations, atomic operations 

## Impact

The driver now works in dcache WB mode  on ETH 1 or 2 (defaults to 1 to avoind breakage.

## Testing

PX4 nxp_fmurt160-v2 and QGC

### CODE REVIEW

File: arch/arm/src/imxrt/imxrt_enet.c [hunk 1]
```diff
@@ -269,34 +314,48 @@ struct imxrt_driver_s
...
+static uint8_t g_buffer_pool[NENET_NBUFFERS][ALIGNED_BUFSIZE]
+                             aligned_data(ENET_ALIGN);
 
 /****************************************************************************
  * Private Function Prototypes
  ****************************************************************************/
 
 /* Utility functions */
 
+static inline uint32_t imxrt_enet_getreg32(FAR struct imxrt_driver_s *priv,
```
Comment (pkarashchenko) [inline arch/arm/src/imxrt/imxrt_enet.c:349]:
```suggestion
static inline uint32_t imxrt_enet_getreg32(struct imxrt_driver_s *priv,
```

File: arch/arm/src/imxrt/imxrt_enet.c [hunk 2]
```diff
@@ -269,34 +314,48 @@ struct imxrt_driver_s
...
+                                           uint32_t offset);
+static inline void imxrt_enet_putreg32(FAR struct imxrt_driver_s *priv,
```
Comment (pkarashchenko) [inline arch/arm/src/imxrt/imxrt_enet.c:351]:
```suggestion
static inline void imxrt_enet_putreg32(struct imxrt_driver_s *priv,
```

File: arch/arm/src/imxrt/imxrt_enet.c [hunk 3]
```diff
@@ -269,34 +314,48 @@ struct imxrt_driver_s
...
+                                       uint32_t value, uint32_t offset);
+
+static inline void imxrt_enet_modifyreg32(FAR struct imxrt_driver_s *priv,
```
Comment (pkarashchenko) [inline arch/arm/src/imxrt/imxrt_enet.c:354]:
```suggestion
static inline void imxrt_enet_modifyreg32(struct imxrt_driver_s *priv,
```

File: arch/arm/src/imxrt/imxrt_enet.c [hunk 4]
```diff
@@ -374,6 +433,74 @@ static void imxrt_reset(struct imxrt_driver_s *priv);
...
+ *   offset    - offset to the register of interest
+ *   clearbits - the 32-bit value to be written as 0s
+ *   setbits   - the 32-bit value to be written as 1s
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static inline void imxrt_enet_modifyreg32(FAR struct imxrt_driver_s *priv,
```
Comment (pkarashchenko) [inline arch/arm/src/imxrt/imxrt_enet.c:474]:
```suggestion
static inline void imxrt_enet_modifyreg32(struct imxrt_driver_s *priv,
```

File: arch/arm/src/imxrt/imxrt_enet.c [hunk 5]
```diff
@@ -374,6 +433,74 @@ static void imxrt_reset(struct imxrt_driver_s *priv);
...
+ *   priv   - private ENET device structure
+ *   offset - offset to the register of interest
+ *
+ * Returned Value:
+ *   The contents of the 32-bit register
+ *
+ ****************************************************************************/
+
+static inline uint32_t
+imxrt_enet_getreg32(FAR struct imxrt_driver_s *priv, uint32_t offset)
```
Comment (pkarashchenko) [inline arch/arm/src/imxrt/imxrt_enet.c:452]:
```suggestion
static inline uint32_t imxrt_enet_getreg32(struct imxrt_driver_s *priv, 
                                           uint32_t offset)
```

File: arch/arm/src/imxrt/imxrt_enet.c [hunk 6]
```diff
@@ -374,6 +433,74 @@ static void imxrt_reset(struct imxrt_driver_s *priv);
...
+ *   priv   - private SPI device structure
+ *   value  - the 32-bit value to be written
+ *   offset - offset to the register of interest
+ *
+ * Returned Value:
+ *   The contents of the 32-bit register
+ *
+ ****************************************************************************/
+
+static inline void imxrt_enet_putreg32(FAR struct imxrt_driver_s *priv,
```
Comment (pkarashchenko) [inline arch/arm/src/imxrt/imxrt_enet.c:498]:
```suggestion
static inline void imxrt_enet_putreg32(struct imxrt_driver_s *priv,
```

File: arch/arm/src/imxrt/imxrt_enet.c [hunk 7]
```diff
@@ -527,41 +652,47 @@ static int imxrt_transmit(struct imxrt_driver_s *priv)
...
+
+  if (rxdesc->data == buf)
+    {
       /* Data was written into the RX buffer, so swap the TX and RX buffers */
 
       DEBUGASSERT((rxdesc->status1 & RXDESC_E) == 0);
       rxdesc->data = txdesc->data;
       txdesc->data = buf;
+      up_clean_dcache((uintptr_t) rxdesc,
+                      (uintptr_t) rxdesc + sizeof(struct enet_desc_s));
```
Comment (pkarashchenko) [inline arch/arm/src/imxrt/imxrt_enet.c:669]:
```suggestion
      up_clean_dcache((uintptr_t)rxdesc,
                      (uintptr_t)rxdesc + sizeof(struct enet_desc_s));
```

File: arch/arm/src/imxrt/imxrt_enet.c [hunk 8]
```diff
@@ -527,41 +652,47 @@ static int imxrt_transmit(struct imxrt_driver_s *priv)
...
     }
   else
     {
       DEBUGASSERT(txdesc->data == buf);
     }
 
-  /* Make the following operations atomic */
+  up_clean_dcache((uintptr_t) txdesc,
+                  (uintptr_t) txdesc + sizeof(struct enet_desc_s));
```
Comment (pkarashchenko) [inline arch/arm/src/imxrt/imxrt_enet.c:677]:
```suggestion
  up_clean_dcache((uintptr_t)txdesc,
                  (uintptr_t)txdesc + sizeof(struct enet_desc_s));
```

File: arch/arm/src/imxrt/imxrt_enet.c [hunk 9]
```diff
@@ -527,41 +652,47 @@ static int imxrt_transmit(struct imxrt_driver_s *priv)
...
 
-  flags = spin_lock_irqsave(NULL);
+  up_clean_dcache((uintptr_t) priv->dev.d_buf,
+                  (uintptr_t) priv->dev.d_buf + priv->dev.d_len);
```
Comment (pkarashchenko) [inline arch/arm/src/imxrt/imxrt_enet.c:680]:
```suggestion
  up_clean_dcache((uintptr_t)priv->dev.d_buf,
                  (uintptr_t)priv->dev.d_buf + priv->dev.d_len);
```

File: arch/arm/src/imxrt/imxrt_enet.c [hunk 10]
```diff
@@ -858,6 +989,9 @@ static void imxrt_receive(struct imxrt_driver_s *priv)
             imxrt_swap32((uint32_t)priv->txdesc[priv->txhead].data);
           rxdesc->status1 |= RXDESC_E;
 
+          up_clean_dcache((uintptr_t) rxdesc,
+                          (uintptr_t) rxdesc + sizeof(struct enet_desc_s));
```
Comment (pkarashchenko) [inline arch/arm/src/imxrt/imxrt_enet.c:993]:
```suggestion
          up_clean_dcache((uintptr_t)rxdesc,
                          (uintptr_t)rxdesc + sizeof(struct enet_desc_s));
```

File: arch/arm/src/imxrt/imxrt_enet.c [hunk 11]
```diff
@@ -1097,14 +1226,15 @@ static void imxrt_enet_interrupt_work(void *arg)
 
 static int imxrt_enet_interrupt(int irq, void *context, void *arg)
 {
-  register struct imxrt_driver_s *priv = &g_enet[0];
+  register FAR struct imxrt_driver_s *priv =
+                                          (FAR struct imxrt_driver_s *) arg;
```
Comment (pkarashchenko) [inline arch/arm/src/imxrt/imxrt_enet.c:1230]:
```suggestion
  register struct imxrt_driver_s *priv =
    (struct imxrt_driver_s *)arg;
```

File: arch/arm/src/imxrt/imxrt_enet.c [hunk 12]
```diff
@@ -2320,14 +2454,17 @@ static void imxrt_initbuffers(struct imxrt_driver_s *priv)
...
+      addr                   += ALIGNED_BUFSIZE;
     }
 
   /* Set the wrap bit in the last descriptors to form a ring */
 
   priv->txdesc[CONFIG_IMXRT_ENET_NTXBUFFERS - 1].status1 |= TXDESC_W;
   priv->rxdesc[CONFIG_IMXRT_ENET_NRXBUFFERS - 1].status1 |= RXDESC_W;
 
+  up_clean_dcache((uintptr_t) g_desc_pool,
+                  (uintptr_t) g_desc_pool + sizeof(g_desc_pool));
```
Comment (pkarashchenko) [inline arch/arm/src/imxrt/imxrt_enet.c:2466]:
```suggestion
  up_clean_dcache((uintptr_t)g_desc_pool,
                  (uintptr_t)g_desc_pool + sizeof(g_desc_pool));
```

**Comment by pkarashchenko** [comment]
`Error: chip/imxrt_clockconfig.c:276:12: error: 'CCM_ANALOG_PLL_ENET_ENABLE' undeclared (first use in this function); did you mean 'CCM_ANALOG_PLL_USB2_ENABLE'?`

**Comment by davids5** [comment]
@pkarashchenko @xiaoxiang781216 - Thank you for the review. I will incorporate all the changes and fix CI - then force push.

File: arch/arm/src/imxrt/imxrt_enet.c [hunk 13]
```diff
@@ -125,6 +125,12 @@
 #  error Write back D-Cache not yet supported
 #endif
 
+/* TX poll delay = 1 seconds. CLK_TCK is the number of clock ticks per
+ * second.
+ */
+
+#define IMXRT_WDDELAY     (1 * CLK_TCK)
```
[the remark this answers is not in the document]
Comment (davids5) [reply arch/arm/src/imxrt/imxrt_enet.c]: This is Odd. Do you see this is the file?

**Comment by pkarashchenko** [comment]
CI still complains on
```
Error: chip/imxrt_enet.c:120:37: error: 'GPR_GPR1_ENET1_TX_DIR_OUT' undeclared (first use in this function); did you mean 'GPR_GPR1_ENET_TX_DIR'?
```

**Comment by davids5** [comment]
> CI still complains on
> 
> ```
> Error: chip/imxrt_enet.c:120:37: error: 'GPR_GPR1_ENET1_TX_DIR_OUT' undeclared (first use in this function); did you mean 'GPR_GPR1_ENET_TX_DIR'?
> ```
Yes - I am tracking it down and have had 3 levels of interrupt.....reti now

File: arch/arm/src/imxrt/imxrt_enet.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply arch/arm/src/imxrt/imxrt_enet.c]: you add in patch:
https://github.com/apache/incubator-nuttx/pull/6787/commits/d56cbc7d296b7f535c563e6cc67644f8b29bda09#diff-17de76e9fae5c5fa9177a5c7c101870dab6729a68d2cd9650491b91926d5fa61R132
and remove here:
https://github.com/apache/incubator-nuttx/pull/6787/commits/b4dddc245d87a657bcd95afbac68a0c53ede1215#diff-17de76e9fae5c5fa9177a5c7c101870dab6729a68d2cd9650491b91926d5fa61L132

[the remark this answers is not in the document]
Comment (davids5) [reply arch/arm/src/imxrt/imxrt_enet.c]: Ahh Thanks It was the merge from the cherry-pick

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by masayuki2009** [comment]
@davids5 
Sorry for adding my comments to this old PR.
In this PR, it seems that you fixed write-back cache issues in imxrt_enet.c. 
So I removed the write-through configuration from imxrt1060-evk:netnsh like

```
--- a/boards/arm/imxrt/imxrt1060-evk/configs/netnsh/defconfig
+++ b/boards/arm/imxrt/imxrt1060-evk/configs/netnsh/defconfig
@@ -14,7 +14,6 @@ CONFIG_ARCH_CHIP_IMXRT=y
 CONFIG_ARCH_CHIP_MIMXRT1062DVL6A=y
 CONFIG_ARCH_STACKDUMP=y
 CONFIG_ARMV7M_DCACHE=y
-CONFIG_ARMV7M_DCACHE_WRITETHROUGH=y
 CONFIG_ARMV7M_ICACHE=y
 CONFIG_ARMV7M_USEBASEPRI=y
 CONFIG_BOARD_LOOPSPERMSEC=104926
```

And tried to see how the networking with imxrt1060-evk is stable.
However, it stopped after several minutes.
Do you think d-cache write-back mode works with imxrt1060-evk?

**Comment by davids5** [comment]
@

> @davids5 Sorry for adding my comments to this old PR. In this PR, it seems that you fixed write-back cache issues in imxrt_enet.c. So I removed the write-through configuration from imxrt1060-evk:netnsh like
> 
> ```
> --- a/boards/arm/imxrt/imxrt1060-evk/configs/netnsh/defconfig
> +++ b/boards/arm/imxrt/imxrt1060-evk/configs/netnsh/defconfig
> @@ -14,7 +14,6 @@ CONFIG_ARCH_CHIP_IMXRT=y
>  CONFIG_ARCH_CHIP_MIMXRT1062DVL6A=y
>  CONFIG_ARCH_STACKDUMP=y
>  CONFIG_ARMV7M_DCACHE=y
> -CONFIG_ARMV7M_DCACHE_WRITETHROUGH=y
>  CONFIG_ARMV7M_ICACHE=y
>  CONFIG_ARMV7M_USEBASEPRI=y
>  CONFIG_BOARD_LOOPSPERMSEC=104926
> ```
> 
> And tried to see how the networking with imxrt1060-evk is stable. However, it stopped after several minutes. Do you think d-cache write-back mode works with imxrt1060-evk?

Yes. I have not tested on current master but it was working here https://github.com/PX4/PX4-Autopilot/tree/pr-fmu1062/boards/nxp/fmurt1062-v2 with NuttX 10.3+

**Pull request merged**